### ISSUE TITLE
webserver crashes if no physical disks are available

### ISSUE THREAD

**Issue by pmav99** [FritsHoogland/procstat#3]
VMs of EuroHPC don't have physical disks. Everything seems to be mounted on tmpfs and persistence is achieved by mounting 2 LustreFS instances:

```
$ df -h
Filesystem                                                                                                                 Size  Used Avail Use% Mounted on
devtmpfs                              252G     0  252G   0% /dev
tmpfs                                 252G   74M  252G   1% /run
tmpfs                                 252G     0  252G   0% /sys/fs/cgroup
tmpfs                                 252G  2.7G  249G   2% /
tmpfs                                 428G  4.4M  428G   1% /dev/shm
10.0.0.0@o2ib,...:/tier1              650T  176T  442T  29% /mnt/tier1
10.1.0.0@o2ib,...:/tier2               12P  405T   11P   4% /mnt/tier2
tmpfs                                  51G     0   51G   0% /run/user/101348
```

Nevertheless, this seems to cause `procstat -w` to crash with:
```
$ ./target/release/procstat -w
Webserver is started at port: 1111
thread 'main' panicked at /mnt/tier2/users/u101348/procstat/src/processor/blockdevice.rs:89:62:
index out of bounds: the len is 0 but the index is 0
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The problem seems to be that `disk_list` array is assumed to contain items:
https://github.com/FritsHoogland/procstat/blob/5cc4237a208ad0f67c6408e7e759f4f00fd5c17e/src/processor/blockdevice.rs#L89

The full backtrace:
```
stack backtrace:
   0:     0x555fafde07f6 - std::backtrace_rs::backtrace::libunwind::trace::h92ba5ac126ecc5ad
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x555fafde07f6 - std::backtrace_rs::backtrace::trace_unsynchronized::hfaa118b6e3a8f31a
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x555fafde07f6 - std::sys_common::backtrace::_print_fmt::h2075f185adf13bbd
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x555fafde07f6 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9c4bd387f9f3f544
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x555fafe0aa10 - core::fmt::rt::Argument::fmt::h018ecebdf268b5bb
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/fmt/rt.rs:142:9
   5:     0x555fafe0aa10 - core::fmt::write::h938c332fdab924eb
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/fmt/mod.rs:1120:17
   6:     0x555fafddd86f - std::io::Write::write_fmt::h4a694b02e44e6363
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/io/mod.rs:1846:15
   7:     0x555fafde05d4 - std::sys_common::backtrace::_print::h3595c0f4303f3ccd
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x555fafde05d4 - std::sys_common::backtrace::print::ha888e6736b0bc71f
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x555fafde1e57 - std::panicking::default_hook::{{closure}}::he19a7f79f7beab5e
  10:     0x555fafde1bb9 - std::panicking::default_hook::h67efe04e9a5d446e
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panicking.rs:292:9
  11:     0x555fafde22e8 - std::panicking::rust_panic_with_hook::h49021cdbc4b22349
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panicking.rs:781:13
  12:     0x555fafde21c2 - std::panicking::begin_panic_handler::{{closure}}::hfbf601f3d8c62d13
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panicking.rs:659:13
  13:     0x555fafde0cf6 - std::sys_common::backtrace::__rust_end_short_backtrace::h98dd020b6e913806
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/sys_common/backtrace.rs:171:18
  14:     0x555fafde1f14 - rust_begin_unwind
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panicking.rs:647:5
  15:     0x555faf7815d5 - core::panicking::panic_fmt::h0d3f1893e38be419
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/panicking.rs:72:14
  16:     0x555faf781812 - core::panicking::panic_bounds_check::hf45ce6debf926a45
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/panicking.rs:208:5
  17:     0x555faf7a2c0b - <usize as core::slice::index::SliceIndex<[T]>>::index::hf72b05af885164b6
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/slice/index.rs:255:10
  18:     0x555faf7a2c0b - core::slice::index::<impl core::ops::index::Index<I> for [T]>::index::hbd028929c8f22981
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/slice/index.rs:18:9
  19:     0x555faf7a2c0b - <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index::hd81dcc09b7edd2f4
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/alloc/src/vec/mod.rs:2771:9
  20:     0x555faf7a2c0b - procstat::processor::blockdevice::add_blockdevices_to_history::{{closure}}::h8d16474ea1f0cdc3
                               at /mnt/tier2/users/user_name/procstat/src/processor/blockdevice.rs:89:62
  21:     0x555faf7ade02 - procstat::processor::add_to_history::{{closure}}::hd3d4818403e85b44
                               at /mnt/tier2/users/user_name/procstat/src/processor/mod.rs:135:45
  22:     0x555faf7ade02 - procstat::processor::read_proc_data_and_process::{{closure}}::h20f02aa8b6d772a8
                               at /mnt/tier2/users/user_name/procstat/src/processor/mod.rs:115:36
  23:     0x555faf829a0d - procstat::app::app::{{closure}}::h0a4f381825cb5a57
                               at /mnt/tier2/users/user_name/procstat/src/app.rs:29:61
  24:     0x555faf829a0d - procstat::main::{{closure}}::h0099cd178cfec259
                               at /mnt/tier2/users/user_name/procstat/src/main.rs:78:11
  25:     0x555faf78354d - tokio::runtime::park::CachedParkThread::block_on::{{closure}}::ha9316fb8a240fbce
                               at /home/users/user_name/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/park.rs:282:63
  26:     0x555faf78354d - tokio::runtime::coop::with_budget::hd5ff2d9441383461
                               at /home/users/user_name/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/coop.rs:107:5
  27:     0x555faf78354d - tokio::runtime::coop::budget::h0b9b9c9c30976276
                               at /home/users/user_name/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/coop.rs:73:5
  28:     0x555faf78354d - tokio::runtime::park::CachedParkThread::block_on::h052dd956a75a02a8
                               at /home/users/user_name/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/park.rs:282:31
  29:     0x555fafa59089 - tokio::runtime::context::blocking::BlockingRegionGuard::block_on::h679bd68a08125ffa
                               at /home/users/user_name/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/context/blocking.rs:66:9
  30:     0x555fafa59089 - tokio::runtime::scheduler::multi_thread::MultiThread::block_on::{{closure}}::h314892f26905db7a
                               at /home/users/user_name/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/scheduler/multi_thread/mod.rs:87:22
  31:     0x555fafa59089 - tokio::runtime::context::runtime::enter_runtime::hd4e02cbf3412a076
                               at /home/users/user_name/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/context/runtime.rs:65:16
  32:     0x555faf9fcb7d - tokio::runtime::scheduler::multi_thread::MultiThread::block_on::hab0e89453ac70650
                               at /home/users/user_name/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/scheduler/multi_thread/mod.rs:86:9
  33:     0x555faf9fcb7d - tokio::runtime::runtime::Runtime::block_on::hb138df7b8ae8a782
                               at /home/users/user_name/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/runtime.rs:350:50
  34:     0x555faf9be0b7 - procstat::main::h34d21efc6919ac10
                               at /mnt/tier2/users/user_name/procstat/src/main.rs:82:5
  35:     0x555faf9c5233 - core::ops::function::FnOnce::call_once::hf1d53eed231a82e7
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/ops/function.rs:250:5
  36:     0x555faf9c5233 - std::sys_common::backtrace::__rust_begin_short_backtrace::ha3ed3b42c47c1de1
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/sys_common/backtrace.rs:155:18
  37:     0x555fafa55bfd - std::rt::lang_start::{{closure}}::h5bf59f2db78153a4
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/rt.rs:166:18
  38:     0x555fafdd4221 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::he2501730d3d7649e
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/ops/function.rs:284:13
  39:     0x555fafdd4221 - std::panicking::try::do_call::h900df580cee6b404
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panicking.rs:554:40
  40:     0x555fafdd4221 - std::panicking::try::h2835d399792b43db
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panicking.rs:518:19
  41:     0x555fafdd4221 - std::panic::catch_unwind::hdff7b6e6c35776df
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panic.rs:142:14
  42:     0x555fafdd4221 - std::rt::lang_start_internal::{{closure}}::h6f12a67e0af11221
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/rt.rs:148:48
  43:     0x555fafdd4221 - std::panicking::try::do_call::h226d8232820746bf
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panicking.rs:554:40
  44:     0x555fafdd4221 - std::panicking::try::h7a70dd2f0620aaac
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panicking.rs:518:19
  45:     0x555fafdd4221 - std::panic::catch_unwind::h4827d3381ec3b157
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panic.rs:142:14
  46:     0x555fafdd4221 - std::rt::lang_start_internal::hc3f700406209db2c
                               at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/rt.rs:148:20
  47:     0x555faf9be19c - main
  48:     0x152ca0d84d85 - __libc_start_main
  49:     0x555faf781e21 - <unknown>
```

**Comment by FritsHoogland** [comment]
Sorry, just seen your issue now. Let me look into it.
Until now, procstat expects at least one physical disk to exist. In your case there is no physical disk, and therefore it crashes on that.

**Comment by FritsHoogland** [comment]
@pmav99 The latest version should at least throw a more human readable error: Unable to find key in hash map statistics, key1: blockdevice, key2: , key3: stat_reads_completed_success.

**Comment by FritsHoogland** [comment]
This is fixed with version 0.1.25
